### ISSUE TITLE
AngularJs 2.0 and future of the Angular-Gantt component

### ISSUE THREAD

**Issue by shansubra** [angular-gantt/angular-gantt#350]
First of all thanks for the awesome component. FYI, I am using this component in my project and it does everything what I needed.

Actually I heard lot of things about AngularJS 2.0 and the immediate thing which caught my eyes were the following features wont be supported in the new AngularJS version:

controller
directive
$scope 
and so on...

 So, what would be future of this Gantt component? I know this is too early to think about this but I am really worried about that after reading few articles.

**Comment by Schweigi** [comment]
This is an good question. I think this will be a general problem as all apps will have to be either ported/rewritten or will stay on Angular 1.0 until their not needed anymore. It is inevitable that the Angular community will take a hit because of this like it already happened to other framework/languages where no backward compatibility was given. But I personally think that progress is a good thing and Angular 2.0 has some really nice features.

Regarding angular-gantt I have to say that the core parts are pure JS so they could be re-used but many things have to be re-written as they are UI specific and this code is not compatible. I also think that a re-writte only makes sense after Angular 2.0 is stable as there will probably be a lot of bugs and issues in the beginning. At this point maybe there will already be another framework which is even better - who knows? :) The JS framework world is fast moving.

**Comment by Toilal** [comment]
:+1: 

I'm not sure I'll use AngularJS 2.0, because it's some much change that it worth considering others framework too. It would be great to port this component to something more "vanilla JavaScript" and framework agnostic, but it's some much work and harder to maintain ... 

Currently AngularJS 1.x is still there, and it's great, so lets keep use it :).

**Comment by kschwidder** [comment]
:+1:

**Comment by Toilal** [comment]
I reopen this issue, because I plan to port angular-gantt to angular 2 ... Development will be in 2.x branch.

**Comment by Schweigi** [comment]
@Toilal Would it make sense to create a angular2-gantt repo instead of a branch? Then it would be easier to track issues, commits and have a clean separation.

**Comment by Toilal** [comment]
Yes for sure. I don't know if there's already some convention to name
angularJS 2 modules and npm/bower package...
Le 6 oct. 2015 18:19, "Marco Schweighauser" <notifications@github.com> a
écrit :

**Comment by khru** [comment]
Would you rewrite the library with Angular 2? and if so, what would you change?

**Comment by Toilal** [comment]
Sadly, I don't have time to perform this migration. If one day I have time, it will be a new project, and everything would change.

**Comment by ryanlin1986** [comment]
Any updates on this?

**Comment by Toilal** [comment]
I started a [new project based on Vue.js](https://github.com/Toilal/vue-gantt), but currently it's still empty as I lack time to work on it.

It seems to make more sense today to build a component on a view only library instead of a full pledged framework. [Vue.js](https://vuejs.org/) is one of the smallest library, so I hope there will soon be some viable solution to import a Vue.js component as an Angular module.

But at my work, we are currently thinking about the right JS Framework to choose between Angular and Vue, so maybe i'll switch to Angular soon ...

**Comment by ryanlin1986** [comment]
Please consider Polymer as well, I think it's very promising.

**Comment by Toilal** [comment]
Also, it seems somebody started an angular 2 component. 

https://github.com/dalestone/angular2-gantt

**Comment by Schweigi** [comment]
I think with ES6 it would be possible to make a good core component and then provide a view layer on top of it using different technologies (e.g. Vue, Angular, React). 

Basically what people are doing by creating JQuery components and then providing a Angular integration. Only in this case the integration also provides the actual html part and the JQuery part is replaced with pure ES6. 

Not sure if it would work - but it could provide a good future proof base.

Btw. Telerik is currently re-creating their components with Angular 2 and not ReactJS, so that's an interesting decision.

**Comment by Toilal** [comment]
Of course, it would be great to support multiple frameworks using a core library in TypeScript transpiled to ES6, but we would need more contributors to support each framework. I think we can choose this architecture but still focus to one framework as the reference implementation. 

It also make things more complex, and we must find the right tooling to make the development environment as easy and straightforward as possible. I'll try to set up this kind of stack on my vuejs template to give it a try.

**Comment by Toilal** [comment]
[bitjson/typescript-starter](https://github.com) seems to be a good boilerplate to start the library part. It allows to write the library in typescript and package it for TypeScript, ES6 and ES5 users.

As we will have multiple npm package to handle, maybe we could use [Lerna](https://lernajs.io/).

**Comment by Toilal** [comment]
I create a new repository for this project: https://github.com/angular-gantt/gantt. Please open issue in this new repository for everything related to next major version of angular gantt.

**Comment by Toilal** [comment]
I reopen because I still have some ideas about this. I just found this blog post which is just brilliant : https://medium.com/@priley86/building-cross-framework-components-with-performance-in-mind-a-web-component-thesis-5e17353aee18#.i3vv8n3j8

The blog post advocate for this idea : If you need to support many framework for a component, vanillaJS is a mess, but you can choose a really minimal DOM library to write it, like [SkateJS](https://github.com/skatejs/skatejs) or [PreactJS](https://github.com/developit/preact).

Also, there's a project called [Onsen.io](https://onsen.io/). It seems to exploit this idea : Core components are written using a minimalist WebComponent implementation, and some bindings libraries allow to use them with any view framework : AngularJS, Angular, React, Vue are currently supported.

**Comment by mklinke** [comment]
@Toilal First of all, we really appreciate your effort to build and maintain this component that we have successfully used in AngularJS projects. As we are currently in the process of building a new app with Angular (4), we have briefly looked into the possibility to use [ngUpgrade](https://angular.io/guide/upgrade#upgrading-with-ngupgrade) to run the angular-gantt component within an Angular application. However, there are a few [constraints](https://angular.io/guide/upgrade#using-component-directives) that need to be fulfilled in order to use ngUpgrade. Unfortunately, the angular-gantt does not fulfill these. Ideally, the component should already be built based on the Component API introduced with AngularJS 1.5.
Do you have a gut feeling about what it would mean to migrate the angular-gantt to the Component API? 

Thanks in advance,
Martin

**Comment by ascode** [comment]
We are maily useing angular4

**Comment by eirikb** [comment]
I think what @mklinke is purposing is the way to go. From what I can tell this project is still by far the best gantt diagram alternative. My dream scenario would be building (usage) support for Webpack without any special configuration, or even Parceljs, and then being able to use it in Angular using ngUpgrade.

**Comment by Toilal** [comment]
@eirikb I'm not sure you are aware there's a [2.0.0-rc.1](https://github.com/angular-gantt/angular-gantt/tree/v2.0.0-rc.1) version that is refactored with TypeScript and "should" work properly with Webpack. (I wrote "should" because I think there are some opened issues related to this version). Consider this as a first step to provide ngUpgrade support, but it still requires to migrate everything to components.

**Comment by eirikb** [comment]
Hi @Toilal , I saw that actually, but I thought it was about building the project, and not usage. 
I will definitely try it out. 
I really appreciate all the work you put into this library. Thanks!

**Comment by Toilal** [comment]
As it's build with Webpack, it outputs an UMD that should be usable with Webpack without any particular configuration.

**Comment by eirikb** [comment]
@Toilal v2.0.0-rc1 doesn't seem to work out-of-the box, firstly because of [`require('ElementQueries')`](https://github.com/angular-gantt/angular-gantt/blob/e6fbce2f2b4bdece2756a41ab1909956cdf3efc1/dist/angular-gantt.js#L10), as already mentioned in #797 . I tested the Webpack demo repo by @hevbevt and it works, but it requires some config, and if I'm not mistaken it also requires `ganttDependencies.js` to be loaded alongside (not bundled). It should be possible to solve this, just not quite sure how...

**Comment by mariohmol** [comment]
Hi! Thanks for this module! I'm doing all in ang2 now and wondering if this will be available in the future for that!!

What i did for now is .. got on native js and port it to ang2
https://github.com/mariohmol/ng-gantt

but there is a lot of features that does not have comparing to angular-gantt..
